### DISCUSSION TITLE
Compatibility with null-ls ?

### DISCUSSION THREAD

**Discussion by clement-buchart** [mfussenegger/nvim-jdtls#137] [Q&A] [answered]
Hello, 

For the project I'm working on, we're using [prettier-java](https://github.com/jhipster/prettier-java) as a formatter.
I've set-up [null-ls](https://github.com/jose-elias-alvarez/null-ls.nvim) as an additional lsp server to handle the formatting. 

Unfortunately nvim-jdtls uses `vim.lsp.buf_request` to execute it's requests to jdtls server, but this method, as per the docs, `Sends an async request for all active clients attached to the buffer.`, including null-ls. 

Obviously null-ls can't handle things like `'vscode.java.resolveMainClass'`, `'vscode.java.test.junit.argument'` and simply returns nothing, and as such using features like `test_nearest_method` throws a lot of errors, and can even crash neovim on my install.

I'm wondering if, instead of `vim.lsp.buf_request`, nvim-jdtls could use something like `vim.lsp.get_active_clients()`, find jdtls, and than use `request(method, params, [handler], bufnr)` on said client. 

If you think that's something reasonable, I would glagly offer a PR with the change.

Cheers.

**Comment by clement-buchart** [comment]
For reference here is the fix I'm using locally : https://github.com/clement-buchart/nvim-jdtls/commit/26d87ce45d8ca4e0eee4c8dc34b97c6034fe6a11 

It would need to handle the client name not being the default value of 'jdt.ls' as it is configurable.
But otherwise it solves my use-case

**Comment by mfussenegger** [comment]
Could you try with https://github.com/mfussenegger/nvim-jdtls/pull/140 ?

Went with a bit of a different approach to not depend on the name. And it might also lead to better error messages if the java-debug bundle isn't setup correctly.

**Comment by clement-buchart** [comment, marked as answer]
Hello, 

Ah yes that looks like a better fix.
The errors are gone on the `command-capabilities` branch.

Thank you.
